File: code/bedrock/07_runtime_config/query.cpp
/*
 * (C) 2024 The University of Chicago
 *
 * See COPYRIGHT in top-level directory.
 */
#include <bedrock/Client.hpp>
#include <bedrock/ServiceHandle.hpp>
#include <nlohmann/json.hpp>
#include <iostream>

int main(int argc, char** argv) {
    if(argc != 2) {
        std::cerr << "Usage: " << argv[0] << " <server_address>" << std::endl;
        return 1;
    }

    try {
        thallium::engine engine("na+sm", THALLIUM_CLIENT_MODE);
        bedrock::Client client(engine);
        bedrock::ServiceHandle service = client.makeServiceHandle(argv[1], 0);

        // Get the complete configuration
        std::string config_str;
        service.getConfig(&config_str);
        nlohmann::json config = nlohmann::json::parse(config_str);

        std::cout << "=== Complete Configuration ===" << std::endl;
        std::cout << config.dump(2) << std::endl;

        // Extract specific information
        std::cout << "\n=== Providers ===" << std::endl;
        if(config.contains("providers")) {
            for(auto& provider : config["providers"]) {
                std::cout << "  - " << provider["name"]
                          << " (type=" << provider["type"]
                          << ", id=" << provider["provider_id"] << ")" << std::endl;
            }
        }

        // Extract pool information
        std::cout << "\n=== Argobots Pools ===" << std::endl;
        if(config.contains("margo") &&
           config["margo"].contains("argobots") &&
           config["margo"]["argobots"].contains("pools")) {
            for(auto& pool : config["margo"]["argobots"]["pools"]) {
                std::cout << "  - " << pool["name"]
                          << " (kind=" << pool["kind"] << ")" << std::endl;
            }
        }

    } catch(const bedrock::Exception& ex) {
        std::cerr << "Error: " << ex.what() << std::endl;
        return 1;
    }

    return 0;
}

File: code/bedrock/05_python/query.py
#!/usr/bin/env python
"""
Example of querying a Bedrock service configuration.
"""
from mochi.bedrock import Client
import sys
import json

if len(sys.argv) != 2:
    print(f"Usage: {sys.argv[0]} <server_address>")
    sys.exit(1)

server_address = sys.argv[1]

# Create client and service handle
client = Client("na+sm")
service = client.make_service_handle(server_address, provider_id=0)

# Get complete configuration
config = service.config
print("=== Complete Configuration ===")
print(json.dumps(config, indent=2))

# Use a Jx9 query to get only provider names
jx9_script = """
$result = [];
foreach($__config__['providers'] as $provider) {
    $result[] = $provider['name'];
}
return $result;
"""

provider_names = service.query(jx9_script)
print("\n=== Provider Names (via Jx9) ===")
print(provider_names)

# Query for pools
jx9_pools = """
$result = [];
if(array_key_exists('margo', $__config__) &&
   array_key_exists('argobots', $__config__['margo']) &&
   array_key_exists('pools', $__config__['margo']['argobots'])) {
    foreach($__config__['margo']['argobots']['pools'] as $pool) {
        $result[] = $pool['name'];
    }
}
return $result;
"""

pools = service.query(jx9_pools)
print("\n=== Argobots Pools ===")
print(pools)

# Clean up
client.finalize()
